Sources/StreamChat/AudioPlayer/AudioPlaying.swift

```diff
@@ -0,0 +1,375 @@
+            guard
+                let self = self,
+                let newValue = newValue
+            else {
+                return
+            }
+
+            let currentPlaybackState = self.context.state
+
+            self.updateContext { value in
+                switch (newValue, currentPlaybackState) {
+                case (.playing, .playing), (.paused, .paused):
+                    break
+                case (.paused, .playing), (.paused, .loading):
+                    value.state = .paused
+                    value.rate = .zero
+                case (.playing, .paused), (.playing, .stopped), (.playing, .loading):
+                    value.state = .playing
+                default:
+                    log.debug("\(type(of: self)): No action for transition \(currentPlaybackState) -> \(newValue)")
+                }
+            }
+        }
+
+        playerObserver.addStoppedPlaybackObserver(
+            queue: nil
+        ) { [weak self] playerItem in
+            guard
+                let self = self,
+                let playerItemURL = (playerItem.asset as? AVURLAsset)?.url,
+                let currentItemURL = (player.currentItem?.asset as? AVURLAsset)?.url,
+                playerItemURL == currentItemURL
+            else {
+                return
+            }
+            var newContext = self.context
+            newContext.state = .stopped
+            newContext.currentTime = 0
+            newContext.rate = .zero
+            newContext.isSeeking = false
+            self.context = newContext
+
+            self.delegate?.audioPlayer(self, didUpdateContext: self.context)
+        }
+    }
+
+    /// Provides thread-safe updates for the player's context and makes sure to forward any updates
+    /// to the the delegate
+    private func updateContext(
+        _ newContextProvider: (inout AudioPlaybackContext) -> Void
+    ) {
+        var newContext = context
+        newContextProvider(&newContext)
+        context = newContext
+        delegate?.audioPlayer(self, didUpdateContext: context)
+    }
+
+    /// It's used by the assetPropertyLoader to handle the completion (successful or failed) of duration's
+    /// asynchronous loading.
+    private func handleDurationLoading(
+        _ result: Result<AVURLAsset, AssetPropertyLoadingCompositeError>
+    ) {
+        guard Thread.isMainThread else {
+            DispatchQueue.main.async { [weak self] in
+                self?.handleDurationLoading(result)
+            }
+            return
+        }
+
+        switch result {
+        /// If the assetPropertyLoaded managed to successfully load the asset's duration information
+        /// we update the context with the new information.
+        case let .success(asset):
+            player.replaceCurrentItem(with: .init(asset: asset))
+            updateContext { value in
+                value.duration = asset.duration.seconds
+                value.currentTime = 0
+                value.rate = .zero
```

Comment:
Shouldn't you update `value.state`? Or you leave it up to `play` to do so?

Reply:
In general I tried to keep update in their related context. For this one where we have basically loaded the asset's properties, we don't have enough context to assume confidently the state of the player. This responsibility is for the timeControlStatusObserver on the player.